File: 13/list.13.1.swift
import Foundation

// ECサイトのユーザモデル
struct User {
  let id: Int
  let name: String
  let email: String
  let passwordDigest: String //ユーザモデル自体にこれいる？
}

final class LoginManager {
  static let shared: LoginManager = .init()
  private init() {}

  private(set) var loginedUser: User?

  func createUser(name: String, email: String, password: String) throws -> Void {
      // 本来はサーバーと通信してユーザ作成
      loginedUser = User(id: 1, name: "", email: "", passwordDigest: "")
  }

  func loginUser(email: String, password: String) throws -> Void {
      // 本来はサーバーと通信してログイン
      loginedUser = User(id: 1, name: "", email: "", passwordDigest: "")
  }

  func delete(user: User) throws -> Void {
      // 本来はサーバーと通信して削除
    loginedUser = nil
  }
}

File: 13/list.13.1.dart
import 'dart:core';

// ECサイトのユーザモデル
class User {
  final int id;
  final String name;
  final String email;
  final String passwordDigest; // ユーザモデル自体にこれいる？

  User(this.id, this.name, this.email, this.passwordDigest);
}

class LoginManager {
  static final LoginManager shared = LoginManager._();
  User? loginedUser;

  LoginManager._();

  Future<void> createUser(String name, String email, String password) async {
    // 本来はサーバーと通信してユーザ作成
    loginedUser = User(1, '', '', '');
  }

  Future<void> loginUser(String email, String password) async {
    // 本来はサーバーと通信してログイン
    loginedUser = User(1, '', '', '');
  }

  Future<void> delete(User user) async {
    // 本来はサーバーと通信して削除
    loginedUser = null;
  }
}
